### ISSUE TITLE
gcovr 5.1 chokes while parsing C++ template

### ISSUE THREAD

**Issue by mtene** [gcovr/gcovr#585]
This is a new issue whn using gcovr 5.1 on our code base, which has not changed recently in this area. Could it be that the const in the template is not recognized by gcovr? A bit confused why it would need to parse the C++ grammar for the purposes of code coverage.

(WARNING) Unrecognized GCOV output for ....
	  784:  Star<const T, ErrorHandler>
	This is indicative of a gcov output parse error.
	Please report this to the gcovr developers
	at <[https://github.com/gcovr/gcovr/issues>.](https://github.com/gcovr/gcovr/issues%3E.)
(WARNING) Exception during parsing:
	UnknownLineType: 784:  Star<const T, ErrorHandler>
(ERROR) Exiting because of parse errors.
	You can run gcovr with --gcov-ignore-parse-errors
	to continue anyway.

**Comment by latk** [comment]
Thank you for reporting this! Could you perhaps provide context of that source code? Which compiler version are you using?

Gcovr does not parse C++ syntax. It parses the .gcov textual coverage reports, but they are sometimes difficult to handle correctly. Normal lines in this .gcov format look like:

```none
       1:   25:  total = 0;
```

that is, three colon-separated fields: execution count, line number, and source code.

The most difficult lines are template specialization sections, which might look like:

```none
------------------
Foo<int>::Foo():
        1:    7:  Foo(): b (1000) {}
------------------
```

Your line `784: Star<const T, ErrorHandler>` is clearly not an ordinary coverage data line because it only has two fields, and clearly not a template specialization section name because it doesn't *end* with a colon.

**Comment by mtene** [comment]
We are using Clang 13 and the line in question is part of an in-header method definition for a templated class:

  template <typename T, typename ErrorHandler = DefaultErrorHandler>
  Star<const T, ErrorHandler>
  wrap_vector(const std::vector<T>& v)
  {
    return ...;
  }

**Comment by latk** [comment]
And that `Star<const T, ErrorHandler>` line is in fact line 784 of the header?

How are you invoking gcovr? In particular, are you using the `--gcov-executable` or `-j` options (or their equivalent configuration file entries)?

Assuming that

* you are using `--gcov-executable 'llvm-cov-13 gcov'` (see [Choosing the right gcov executable](https://gcovr.com/en/stable/guide/compiling.html#choosing-the-right-gcov-executable))
* and are *not* using `-j`

it would be helpful to do the following:

* install the gcovr development version (`pip install git+https://github.com/gcovr/gcovr.git`) – contains some fixes and better error messages for debugging
* run gcovr with the `--verbose` flag
* provide the output around that error, starting with the preceding `(DEBUG) Running gcov: ...` message, possibly redacted
* provide the .gcov file that caused this parse error (see preceding `(DEBUG) Parsing coverage data for file ...` message), possibly redacted

**Comment by latk** [comment]
> `gcovr -j 16`

Then it's likely that this is related to #587, #588, and fixed in the development version with #592. It can also be fixed by removing the `-j` option since it rarely leads to a noticeable speedup (depends entirely on project directory layout + build system).

**Comment by latk** [comment]
I'm closing this assuming that the issue has been resolved. If the issue persists without `-j` or when using the development version, please let me know and provide the information mentioned in the [previous comment](https://github.com/gcovr/gcovr/issues/585#issuecomment-1084479595) so that the issue can be re-opened.
